Core/BookmarksCoreDataStorage.swift

```diff
@@ -422,18 +422,34 @@ extension BookmarksCoreDataStorage {
         loadStore()
     }
     
-    public func favoritesUncachedForWidget(completion: @escaping ([BookmarkManagedObject]) -> Void) {
-        getTopLevelFolder(isFavorite: true, onContext: viewContext) { folder in
-            
-            let children = folder.children?.array as? [BookmarkItemManagedObject] ?? []
-            let favorites: [BookmarkManagedObject] = children.map {
-                if let fav = $0 as? BookmarkManagedObject {
-                    return fav
-                } else {
-                    fatalError("Favourites shouldn't contain folders")
+    public func favoritesUncachedForWidget(completion: @escaping ([BookmarkManagedObject]) -> Void) {        
+        Task {
+            guard await hasTopLevelFolder() else {
+                completion([])
+                return
+            }
+            
+            getTopLevelFolder(isFavorite: true, onContext: viewContext) { folder in
+                let children = folder.children?.array as? [BookmarkItemManagedObject] ?? []
+                let favorites: [BookmarkManagedObject] = children.map {
+                    if let fav = $0 as? BookmarkManagedObject {
+                        return fav
+                    } else {
+                        fatalError("Favourites shouldn't contain folders")
+                    }
                 }
+                completion(favorites)
+            }
+        }
+    }
+    
+    private func hasTopLevelFolder() async -> Bool {
```

Comment:
This isn't the cleanest way to detect whether users have migrated, but I think it's the only choice:

1. We can't check `migratedFromUserDefaults` as it's not stored in a user defaults suite that is accessible to the widget
2. We can't check whether favourites data has been removed from the bookmarks user defaults suite as we're setting that to an empty array post-migration instead of removing it entirely; no way to tell if users are on an older version and simply don't have any favourites, or if they have no favourites because of the migration

Reply:
I think this is the best we can do. Love the use of async!